### PR TITLE
Flesh out Images integration testing 

### PR DESCRIPTION
**Pull request by jwhonce**
* Update Makefile with integration target
* Update client to support unix:// and http+unix:// schemes
* Add Images integration tests
* Refactor PodmanLauncher to log service output to log file to capture debugging events.
* Add option to allow podman service to log debugging events during integration test
* Refactor Image and ImagesManager as needed now integration tests running.
* Refactor unittests where as-built differs from swagger

Included from https://github.com/containers/podman-py/pull/48:
> This change adds a base integration class and a podman launcher that can
> be used to test the api calls and their expected results. Currently the
> podman calls that will work should be ones that can be performed as
> rootless. Eventually calls that require privilege (e.g. network) can be
> implemented but skipped if the tests are not invoked as root.  It should
> be noted that all integration tests will be skipped if a podman
> executable is not in the path.

Signed-off-by: Alex Schultz <aschultz@redhat.com>
Signed-off-by: Jhon Honce <jhonce@redhat.com>

### CODE REVIEW

**Comment by jwhonce** [comment]
@mwhahaha If you have no objections, I've included your commit and built on top of it in this PR. Once I have build() included in the tests I would like to merge this PR and close https://github.com/containers/podman-py/pull/48

**Comment by cevich** [comment]
Looking...I may be a bit slow, since it's all new code to me...

File: podman/domain/images.py [hunk 1]
```diff
@@ -46,7 +52,7 @@ def history(self) -> List[Dict[str, Any]]:
         raise APIError(body["cause"], response=response, explanation=body["message"])
 
     def save(
-        self, chunk_size: Optional[int] = 2097152, named: Union[str, bool] = False
+        self, chunk_size: Optional[int] = 2 * 1024 * 1024, named: Union[str, bool] = False
```
Comment (cevich) [inline podman/domain/images.py:55]: Thanks for spelling out the value-math, it really helps people understand where the value came from.

File: podman/domain/images_manager.py [hunk 2]
```diff
@@ -270,7 +265,6 @@ def pull(
                 config for this request. auth_config should contain the username and password
                 keys to be valid.
             platform (str) – Platform in the format os[/arch[/variant]]
-            all_tags (bool) – Pull all image tags
```
Comment (cevich) [inline podman/domain/images_manager.py:273]: Was this intentional?  I see this is still a keyword arg in the method signature.

File: podman/domain/images_manager.py [hunk 3]
```diff
@@ -306,22 +300,25 @@ def pull(
...
+            body = response.json()
             raise APIError(response.url, response=response, explanation=body["error"])
 
-        if not all_tags:
-            return self.get(body["images"][0])
+        for item in response.iter_lines():
+            body = json.loads(item)
+            if all_tags and "images" in body:
+                images: List[Image] = []
+                for name in body["images"]:
```
Comment (cevich) [inline podman/domain/images_manager.py:315]: Is `body["images"]` guaranteed to always exist in the JSON and always be a list?  If it's a string or some other iteratable, the loop will still function but any error eventually given to the user may not make much sense.  Maybe it's worth type-checking and throwing an exception.  This will help direct any debugging effort closer to the actual problem area (JSON input processing) here.

File: podman/tests/integration/base.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+class IntegrationTest(fixtures.TestWithFixtures):
+    """Base Integration test case"""
+
+    def setUp(self):
+        super().setUp()
+
+        command = os.environ.get("PODMAN_BINARY", "podman")
+        if shutil.which(command) is None:
+            self.skipTest(f"{command} is not installed, skipping integration tests")
```
Comment (cevich) [inline podman/tests/integration/base.py:33]: I'm assuming a test-skip will ultimately result in a *PASS*.  Just as a warning: This can lead to a out-of-sight-out-of-mind situation at a high-level.  For example, PR's getting merged because "The tests are all green" where in reality, they're being skipped and nobody looked.  Granted may be a a corner-case, but it might be worth considering having this fail instead of skip.  If higher-layers shouldn't be running integration tests, they can exclude it at the harness-level more obviously/explicitly, possibly limiting surprises.

File: podman/tests/integration/utils.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        """start podman service"""
+        print(f"Launching {' '.join(self.cmd)}\n\tLogging to {self.log_file.name}")
+        self.proc = subprocess.Popen(self.cmd, stdout=self.log_file, stderr=self.log_file)
+
+        # wait for socket to be created
+        timeout = time.monotonic() + 30
+        while not os.path.exists(self.socket_file):
+            if time.monotonic() > timeout:
+                raise subprocess.TimeoutExpired("podman service ", timeout)
+            time.sleep(0.2)
```
Comment (cevich) [inline podman/tests/integration/utils.py:82]: Is there some positive indicator that could be waited upon instead of sleeping?  Two tenths could either be an eternity on a busy CI system, potentially obscuring another/worse problem.

**Review by cevich — COMMENTED** [review]
Okay, took a more close gander through the changes here.  It's a good start.

File: podman/tests/integration/utils.py (hunk 5 above)
Comment (mwhahaha) [reply podman/tests/integration/utils.py:82]: It's waiting for the socket file to be available. I found that even on local systems a bit of delay occurred prior to the service being ready to accept connections.  While you could lower the time to like 0.01, you definitely don't want to let it spin uncontrolled as it'll just eat CPU.

File: podman/tests/integration/utils.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                f"--time={timeout}",
+                socket_uri,
+            ]
+        )
+
+        self.log_file = tempfile.NamedTemporaryFile(
+            prefix="podman_integration_", suffix=".log", delete=False
+        )
+        self.log_file.write(f"Podman command {' '.join(self.cmd)}\n".encode())
+        self.log_file.flush()
```
[the remark this answers is not in the document]
Comment (mwhahaha) [reply podman/tests/integration/utils.py:70]: This really should use the logging module rather than creating a file like this.

File: podman/domain/image_build.py [hunk 7]
```diff
@@ -155,3 +158,18 @@ def build(self, **kwargs) -> Tuple[Image, Iterator[bytes]]:
...
+            if isinstance(dockerfile, io.BytesIO):
+                info = tarfile.TarInfo("Dockerfile")
+                info.size = len(dockerfile.getvalue())
+                dockerfile.seek(0)
+            else:
+                info = tarball.gettarinfo(fileobj=dockerfile, arcname='Dockerfile')
+            tarball.add(info, dockerfile)
+
+        file.seek(0)
+        return file
```
Comment (TomSweeneyRedHat) [inline podman/domain/image_build.py:175]: PITA time, can we mostly convert Dockerfile to Containerfile throughout?  Perhaps not in this PR, but at some point in the not too distant future?  Not only messaging but also variables within?  Or perhaps make it secondary to Containerfile?   (Tom quickly digs fox hole to hide from Jhon in :bomb: )

File: podman/domain/images.py [hunk 8]
```diff
@@ -60,10 +66,15 @@ def save(
 
         Raises:
             APIError: when service returns an error.
+
+        Notes:
+            tarball format is set to docker-archive, this allows load() to import tarball.
```
Comment (TomSweeneyRedHat) [inline podman/domain/images.py:71]:
```suggestion
            tarball format is set to docker-archive, this allows load() to import a tarball.
```

File: podman/tests/integration/utils.py (hunk 5 above)
Comment (TomSweeneyRedHat) [reply podman/tests/integration/utils.py:82]: Should we have a max tries that would fail the test after 3 or so tries?

Comment (mwhahaha) [reply podman/tests/integration/utils.py:82]: I wouldn't put a loop cap for the exact reasons cevich mentioned, in CI this could be a while so if you looped 3 times you'd only have waited for ~.6 seconds.  There is a cap here of 30 seconds with the timeout var.  30 seconds should be plenty of time to wait for the podman api instance to become available for the integration tests.

File: podman/domain/images_manager.py (hunk 3 above)
Comment (jwhonce) [reply podman/domain/images_manager.py:315]: The check `"images" in body` ensures the key exists, the get() below will fail if the type isn't a list.

File: podman/tests/integration/utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jwhonce) [reply podman/tests/integration/utils.py:70]: I will before duplicating, currently this is a one off.

File: podman/tests/integration/utils.py (hunk 5 above)
Comment (jwhonce) [reply podman/tests/integration/utils.py:82]: @mwhahaha is waiting on the socket to be created on the filesystem. I added the timeout so we don't wait forever and the small sleep should help with CPU usage.

File: podman/domain/image_build.py (hunk 7 above)
Comment (jwhonce) [reply podman/domain/image_build.py:175]: I can change variables and default values, I cannot change the keyword arguments as that would break developers expectations.

File: podman/domain/images_manager.py (hunk 2 above)
Comment (jwhonce) [reply podman/domain/images_manager.py:273]: Nope. Deleted.

File: podman/tests/integration/utils.py (hunk 5 above)
Comment (cevich) [reply podman/tests/integration/utils.py:81]: Okay, I think I understand what's happening here.  The sleep is to prevent checking for the existence of the file, from happening repeatedly as fast as possible.  I don't see it in the code (I didn't look) but a max of 30-seconds for this loop is probably fine.

File: podman/config.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+class PodmanConfig:
+    """PodmanConfig provides a representation of the containers.conf file."""
+
+    def __init__(self, path: str = None):
+        """Read Podman configuration from users XDG_CONFIG_HOME."""
+        # Python 3.6 does not include cached_property therefore must work is done in __init__.
+
+        if path is None:
+            path = os.environ.get("XDG_CONFIG_HOME", f"{Path.home()}/.config/containers")
```
Comment (cevich) [inline podman/config.py:49]: I think this might be wrong.  Did you intend something like this:

```python
path = Path(os.environ.get("XDG_CONFIG_HOME", f"{Path.home()}/.config"))
path = path.joinpath("containers")
```

Comment (jwhonce) [reply podman/config.py:49]: Good catch, next push has no platform dependencies:

```
            home = Path(os.environ.get("XDG_CONFIG_HOME", Path.home().name))
            self.path = home / ".config" / "containers" / "containers.conf"
```

File: podman/tests/integration/utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jwhonce) [reply podman/tests/integration/utils.py:69]: @mwhahaha I went with the file because capturing stderr/stdout via logging conflicted with unittest capturing stdout/stderr. I'll keep playing with it for a future PR.

[the remark this answers is not in the document]
Comment (mwhahaha) [reply podman/tests/integration/utils.py:69]: K. I think there are ways around that and we might just need to include a test friendly logging configuration

**Review by mwhahaha — APPROVED** [review]
LGTM

**Comment by TomSweeneyRedHat** [comment]
LGTM
